ouroboros-consensus/src/Ouroboros/Consensus/Node/Recovery.hs

```diff
@@ -0,0 +1,141 @@
+  => HasFS m h
+  -> m Bool
+hasCleanShutdownMarker hasFS =
+    doesFileExist hasFS cleanShutdownMarkerFile
+
+-- | Create the 'cleanShutdownMarkerFile'.
+--
+-- Will throw an 'FsResourceAlreadyExist' error when it already exists.
+createCleanShutdownMarker
+  :: IOLike m
+  => HasFS m h
+  -> m ()
+createCleanShutdownMarker hasFS = do
+    alreadyExists <- hasCleanShutdownMarker hasFS
+    unless alreadyExists $
+      withFile hasFS cleanShutdownMarkerFile (WriteMode MustBeNew) $ \_h ->
+        return ()
+
+-- | Remove 'cleanShutdownMarkerFile'.
+--
+-- Will throw an 'FsResourceDoesNotExist' error when it does not exist.
+removeCleanShutdownMarker
+  :: IOLike m
+  => HasFS m h
+  -> m ()
+removeCleanShutdownMarker hasFS =
+    removeFile hasFS cleanShutdownMarkerFile
+
+-- | Return 'True' if the given exception indicates that recovery of the
+-- database is required on the next startup.
+exceptionRequiresRecovery :: SomeException -> Bool
+exceptionRequiresRecovery e
+    | Just (ExceptionInLinkedThread _ e') <- fromException e
+    = exceptionRequiresRecovery e'
+    | Just (e' :: ChainDbFailure) <- fromException e
+    = case e' of
+        ImmDbFailure ue -> immDbUnexpectedError ue
+        VolDbFailure ue -> volDbUnexpectedError ue
+        LgrDbFailure fe -> fsError fe
+        _               -> recoveryRequired
+
+    -- The three exceptions below will always be wrapped in a
+    -- 'ChainDbFailure', but we include them just in case.
+    | Just (e' :: ImmutableDBError) <- fromException e
+    = case e' of
+        ImmDB.UnexpectedError ue -> immDbUnexpectedError ue
+        _                        -> noRecoveryRequired
+    | Just (e' :: VolatileDBError) <- fromException e
+    = case e' of
+        VolDB.UnexpectedError ue -> volDbUnexpectedError ue
+        _                        -> noRecoveryRequired
+    | Just (e' :: FsError) <- fromException e
+    = fsError e'
+
+    | otherwise
+    = noRecoveryRequired
+  where
+    recoveryRequired   = True
+    noRecoveryRequired = False
+
+    immDbUnexpectedError :: ImmDB.UnexpectedError -> Bool
+    immDbUnexpectedError = \case
+      ImmDB.FileSystemError fe -> fsError fe
+      _                        -> recoveryRequired
+
+    volDbUnexpectedError :: VolDB.UnexpectedError -> Bool
+    volDbUnexpectedError = \case
+      VolDB.FileSystemError fe -> fsError fe
+      _                        -> recoveryRequired
+
+    fsError :: FsError -> Bool
+    fsError FsError { fsErrorType } = case fsErrorType of
+      FsDeviceFull              -> noRecoveryRequired
+      FsInsufficientPermissions -> noRecoveryRequired
+      _                         -> recoveryRequired
+
+{-------------------------------------------------------------------------------
+  Auxiliary
+-------------------------------------------------------------------------------}
+
+onExceptionIf
+  :: (IOLike m, Exception e)
+  => (e -> Bool)  -- ^ Predicate to selection exceptions
+  -> m b          -- ^ Exception handler
```

Comment:
I never really understand why this customary. Wouldn't `m ()` make a lot more sense? (No need to change it.)

Reply:
Yeah, convenience (no `void`)? Or to make clear from the types which `m X` is the exception handler?

Reply:
I don't like it though. No `void` for the caller, sure, _because we'll implicitly through your result away_. Even `ghc` warns when you do that (in a `do` block). I prefer if it's the caller explicit choice to throw results away -- do you really want an exception handler with a result?